osvr/RenderKit/RenderManagerBase.cpp

```diff
@@ -1704,18 +1737,179 @@ namespace renderkit {
 
             /// Compute the full matrix by multiplying the parts.
             Eigen::Projective3f full =
-                postTranslation * postScale * postProjectionTranslate *
+                postTranslation * justInTimeWarp * postScale * postProjectionTranslate *
                 lastModelViewTransform * currentModelViewInverseTransform *
                 preProjectionTranslate * preScale * preTranslation;
 
-            // Store the result.
+            // Store the result, transposing if we're using D3D.
             matrix16 timeWarp;
-            Eigen::Matrix4f::Map(timeWarp.data) = full.matrix();
+            if (doTranspose) {
+              Eigen::Matrix4f::Map(timeWarp.data) = full.matrix().transpose();
+            } else {
+              Eigen::Matrix4f::Map(timeWarp.data) = full.matrix();
+            }
             m_asynchronousTimeWarps.push_back(timeWarp);
         }
         return true;
     }
 
+    std::array<float, 4> RenderManager::ComputeJustInTimeWarp() {
+      // We initialize the values with ones that won't cause any
+      // change.  We will override them as we find reason.
+      std::array<float, 4> ret = { 1, 1, 0, 0 };
+
+      // Figure out which edge of the display scan-out starts at based
+      // on the just-in-time rotation.  This describes which edge will
+      // be rotated to point "up" when the display is rotated about the
+      // +Z axis (out of the screen) and it starts at the canonical orientation
+      // with X to the right and Y up.  The four results are 0 = top, 1 = right,
+      // 2 = bottom, 3 = left.  The code rounds to the nearest one.
+      int edgeUp = static_cast<int>(
+        floor((m_params.m_justInTimeWarpRotation + 44.9999) / 90));
+      if (edgeUp < 0) { edgeUp += 4 * static_cast<int>(1 - edgeUp / 4); }
+      edgeUp = edgeUp % 4;
+
+      // Find out the timing information, which will let us know the
+      // duration of a full-screen scan-out.  If we are scanning out
+      // from left to right or right to left, divide this by the number
+      // of eyes per display to find the per-eye scan-out duration.
+      RenderTimingInfo timing;
+      if (!GetTimingInfo(0, timing)) {
+        // If we have no timing information, then we have nothing to use
+        // to predict so we return the do-nothing result.
+        return ret;
+      }
+      double screenScanTime = (timing.hardwareDisplayInterval.seconds +
+        timing.hardwareDisplayInterval.microseconds / 1e6);
```

Comment:
1e6 is a double, then the multiplication happens and turns the microseconds into a double, then the sum turns the seconds into a double.  Also working in practice.